File: js/src/containers/ts.js
// MPEG-TS packet and PES walking. spec/10-containers.md §10.1
// Mirrors python/src/dashgps/containers/ts.py

const SYNC = 0x47;
const STRIDES = [188, 192, 204];
const MIN_HITS = 3;
const CONFIRM_HITS = 20;

// Dashcams routinely omit their GPS program from the PMT, so we scan PIDs directly. spec 10.1
export const WELL_KNOWN_PIDS = [0x0300, 0x0102, 0x01e4, 0x0e1b];

const NO_SYNTAX = new Set([0xbc, 0xbe, 0xbf, 0xf0, 0xf1, 0xf2, 0xf8, 0xff]);

export async function detectAlignment(reader, limit = 65536) {
  const n = reader.size();
  const head = await reader.readRange(0, limit < n ? limit : n);
  let best = [0, 0, 0];
  for (const stride of STRIDES) {
    for (let off = 0; off < stride; off++) {
      if (off >= head.length) break;
      if (head[off] !== SYNC) continue;
      const avail = Math.ceil((head.length - off) / stride);
      const want = avail > CONFIRM_HITS ? CONFIRM_HITS : avail;
      let hits = 0;
      for (let p = off; p < head.length; p += stride) {
        if (head[p] !== SYNC) break;
        hits += 1;
      }
      if (hits >= MIN_HITS && hits >= want && hits > best[2]) {
        best = [stride, off, hits];
        if (hits >= CONFIRM_HITS) return [stride, off];
      }
    }
  }
  return best[0] ? [best[0], best[1]] : [0, 0];
}

export async function* iterPackets(reader, stride, offset, pids, chunk = 4 * 1024 * 1024, end) {
  const n = end === undefined ? reader.size() : end;
  const per = Math.max(1, Math.floor(chunk / stride));
  const step = per * stride;
  const want = new Set(pids);
  let pos = offset;
  while (pos + stride <= n) {
    let stop = pos + step;
    if (stop > n) stop = n;
    const buf = await reader.readRange(pos, stop);
    for (let i = 0; i + stride <= buf.length; i += stride) {
      if (buf[i] !== SYNC) continue;
      const b1 = buf[i + 1];
      const pid = ((b1 & 0x1f) << 8) | buf[i + 2];
      if (!want.has(pid)) continue;
      const b3 = buf[i + 3];
      if (!(b3 & 0x10)) continue;
      let j = i + 4;
      if (b3 & 0x20) j += 1 + buf[j];
      if (j < i + 188) yield [pid, (b1 & 0x40) !== 0, buf.subarray(j, i + 188)];
    }
    if (stop >= n) break;
    const advanced = Math.floor(buf.length / stride) * stride;
    pos = advanced > 0 ? pos + advanced : stop;
  }
}

export function stripPesHeader(payload) {
  if (payload.length >= 6 && payload[0] === 0 && payload[1] === 0 && payload[2] === 1) {
    const sid = payload[3];
    if (NO_SYNTAX.has(sid)) return payload.subarray(6);
    if (payload.length >= 9) return payload.subarray(9 + payload[8]);
    return new Uint8Array(0);
  }
  return payload;
}

function concat(chunks, total) {
  const out = new Uint8Array(total);
  let o = 0;
  for (const c of chunks) { out.set(c, o); o += c.length; }
  return out;
}

// Some cameras never set payload_unit_start_indicator on their GPS PID, so payloads accumulate
// regardless when tolerateNoPusi is set. spec 10.1
export async function* iterPes(reader, stride, offset, pids, chunk = 4 * 1024 * 1024,
                               tolerateNoPusi = true, cap = 65536, end) {
  const bufs = new Map();
  const lens = new Map();
  const seen = new Map();
  for await (const [pid, pusi, payload] of
      iterPackets(reader, stride, offset, pids, chunk, end)) {
    if (pusi) {
      seen.set(pid, true);
      const prev = bufs.get(pid);
      if (prev && prev.length) yield [pid, stripPesHeader(concat(prev, lens.get(pid)))];
      bufs.set(pid, [payload]);
      lens.set(pid, payload.length);
    } else {
      let cur = bufs.get(pid);
      if (cur === undefined) {
        if (!tolerateNoPusi && !seen.get(pid)) continue;
        cur = [];
        bufs.set(pid, cur);
        lens.set(pid, 0);
      }
      if (lens.get(pid) < cap) {
        cur.push(payload);
        lens.set(pid, lens.get(pid) + payload.length);
      }
    }
  }
  for (const [pid, chunks] of bufs) {
    if (chunks.length) yield [pid, stripPesHeader(concat(chunks, lens.get(pid)))];
  }
}

File: python/src/dashgps/containers/ts.py
"""MPEG-TS packet and PES walking. spec/10-containers.md §10.1."""

from ..io import scan_chunks

SYNC = 0x47
STRIDES = (188, 192, 204)
MIN_HITS = 3
CONFIRM_HITS = 20

# PIDs dashcams are known to use for GPS. We do not gate on the PMT: these cameras routinely
# omit the GPS program from it entirely. spec 10.1.
WELL_KNOWN_PIDS = (0x0300, 0x0102, 0x01E4, 0x0E1B)

_NO_SYNTAX = frozenset((0xBC, 0xBE, 0xBF, 0xF0, 0xF1, 0xF2, 0xF8, 0xFF))


def detect_alignment(reader, limit=65536):
    """Return (stride, offset) or (0, 0) if this is not a TS stream."""
    n = reader.size()
    head = reader.read_range(0, limit if limit < n else n)
    best = (0, 0, 0)
    for stride in STRIDES:
        for off in range(0, stride):
            if off >= len(head):
                break
            if head[off] != SYNC:
                continue
            avail = (len(head) - off + stride - 1) // stride
            want = CONFIRM_HITS if avail > CONFIRM_HITS else avail
            hits = 0
            p = off
            while p < len(head):
                if head[p] != SYNC:
                    break
                hits += 1
                p += stride
            if hits >= MIN_HITS and hits >= want and hits > best[2]:
                best = (stride, off, hits)
                if hits >= CONFIRM_HITS:
                    return stride, off
    return (best[0], best[1]) if best[0] else (0, 0)


def iter_packets(reader, stride, offset, pids, chunk=4 * 1024 * 1024, end=None):
    """Yield (pid, pusi, payload) for packets whose PID is in ``pids``.

    Reads in chunks aligned to the packet stride so a packet is never split across two reads.
    """
    n = reader.size() if end is None else end
    per = chunk // stride
    if per < 1:
        per = 1
    step = per * stride
    pos = offset
    while pos + stride <= n:
        stop = pos + step
        if stop > n:
            stop = n
        buf = reader.read_range(pos, stop)
        i = 0
        while i + stride <= len(buf):
            if buf[i] != SYNC:
                i += stride
                continue
            b1 = buf[i + 1]
            b2 = buf[i + 2]
            pid = ((b1 & 0x1F) << 8) | b2
            if pid in pids:
                b3 = buf[i + 3]
                if b3 & 0x10:  # has payload
                    j = i + 4
                    if b3 & 0x20:  # adaptation field
                        j += 1 + buf[j]
                    if j < i + 188:
                        yield pid, bool(b1 & 0x40), buf[j : i + 188]
            i += stride
        if stop >= n:
            break
        pos = pos + (len(buf) // stride) * stride
        if pos >= stop:
            pos = stop


def strip_pes_header(payload):
    """Return the payload after any PES header. spec 10.1."""
    if len(payload) >= 6 and payload[0] == 0 and payload[1] == 0 and payload[2] == 1:
        sid = payload[3]
        if sid in _NO_SYNTAX:
            return payload[6:]
        if len(payload) >= 9:
            return payload[9 + payload[8] :]
        return b""
    return payload


def iter_pes(reader, stride, offset, pids, chunk=4 * 1024 * 1024, tolerate_no_pusi=True,
             cap=65536, end=None):
    """Assemble and yield PES payloads per PID.

    Some dashcams never set payload_unit_start_indicator on their GPS PID, so when
    ``tolerate_no_pusi`` is set the payload simply accumulates until the cap. spec 10.1.
    """
    bufs = {}
    seen_pusi = {}
    for pid, pusi, payload in iter_packets(reader, stride, offset, pids, chunk, end):
        if pusi:
            seen_pusi[pid] = True
            prev = bufs.get(pid)
            if prev:
                yield pid, strip_pes_header(bytes(prev))
            bufs[pid] = bytearray(payload)
        else:
            cur = bufs.get(pid)
            if cur is None:
                if not tolerate_no_pusi and not seen_pusi.get(pid):
                    continue
                cur = bytearray()
                bufs[pid] = cur
            if len(cur) < cap:
                cur.extend(payload)
    for pid, buf in bufs.items():
        if buf:
            yield pid, strip_pes_header(bytes(buf))


def iter_raw_chunks(reader, opts, start=0, end=None):
    """Overlapping raw slabs, for scanners that do not care about packet framing."""
    return scan_chunks(reader, opts.chunk, opts.overlap, start, end)
